### PR TITLE
JSON: Show package.json dependency tooltips on hover

### LINKED ISSUE #13303: Show package latest version, link and short description
### Check for existing issues

- [X] Completed

### Describe the feature

It would be good to have the popup shown on hover over the package version to show the package latest version and short description as well as link just like vscode.

### If applicable, add mockups / screenshots to help present your vision of the feature

Not sure whether to create extension or provide built-in.

### PR DESCRIPTION
**Pull request by osiewicz**
Fixes https://github.com/zed-industries/zed/issues/13303

Release Notes:

- Added package version tooltips when hovering over package.json dependency entries.

### CODE REVIEW

File: crates/languages/src/json.rs [hunk 1]
```diff
@@ -244,3 +251,123 @@ fn schema_file_match(path: &Path) -> &Path {
...
+            delegate.http_client(),
+        )
+        .await?;
+        let os = match consts::OS {
+            "macos" => "apple-darwin",
+            "linux" => "unknown-linux-gnu",
+            "windows" => "pc-windows-msvc",
+            other => bail!("Running on unsupported os: {other}"),
+        };
+        let asset_name = format!("package-version-server-{}-{os}.tar.gz", consts::ARCH);
```
Comment (JunkuiZhang) [inline crates/languages/src/json.rs:280]: On windows, I guess this is `.zip` file?
Screenshot 2024-06-25 161749

Comment (osiewicz) [reply crates/languages/src/json.rs]: @JunkuiZhang Yup, I've fixed that up. I have no way of testing this though, as I don't have Zed set up on my Windows machine.

**Comment by notpeter** [comment]
@osiewicz I tried to use this in the new nightly and it's failing with following:
<details><summary>zed.log</summary>
<pre>
2024-06-26T11:51:59-04:00 [INFO] Opening main db
2024-06-26T11:51:59-04:00 [INFO] Initializing default prettier with plugins {}
2024-06-26T11:51:59-04:00 [WARN] Skipping path "/Users/peter/source/robopaint" that has no prettier dependency and no workspaces section in its package.json
2024-06-26T11:51:59-04:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-06-26T11:51:59-04:00 [INFO] starting language server "json-language-server", path: "/Users/peter/source/robopaint", id: 1
2024-06-26T11:51:59-04:00 [INFO] starting language server "package-version-server", path: "/Users/peter/source/robopaint", id: 2
2024-06-26T11:51:59-04:00 [INFO] Initializing default prettier with plugins {}
2024-06-26T11:51:59-04:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-06-26T11:51:59-04:00 [WARN] Skipping path "/Users/peter/source/robopaint" that has no prettier dependency and no workspaces section in its package.json
2024-06-26T11:51:59-04:00 [INFO] fetching latest version of language server "json-language-server"
2024-06-26T11:51:59-04:00 [INFO] Node runtime install_if_needed
2024-06-26T11:51:59-04:00 [INFO] fetching latest version of language server "package-version-server"
2024-06-26T11:52:00-04:00 [INFO] downloading language server "package-version-server"
2024-06-26T11:52:00-04:00 [INFO] downloading language server "json-language-server"
2024-06-26T11:52:00-04:00 [INFO] Node runtime install_if_needed
2024-06-26T11:52:00-04:00 [INFO] starting language server. binary path: "/Users/peter/Library/Application Support/Zed/node/node-v18.15.0-darwin-arm64/bin/node", working directory: "/Users/peter/source/robopaint", args: ["/Users/peter/Library/Application Support/Zed/languages/json-language-server/node_modules/vscode-langservers-extracted/bin/vscode-json-language-server", "--stdio"]
2024-06-26T11:52:00-04:00 [INFO] starting language server. binary path: "/Users/peter/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.1/package-version-server", working directory: "/Users/peter/source/robopaint", args: []
2024-06-26T11:52:00-04:00 [ERROR] failed to start language server "package-version-server": failed to spawn command. path: "/Users/peter/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.1/package-version-server", working directory: "/Users/peter/source/robopaint", args: []
2024-06-26T11:52:00-04:00 [ERROR] server stderr: Some("")
2024-06-26T11:52:00-04:00 [INFO] retrying installation of language server "package-version-server" in 1s
2024-06-26T11:52:01-04:00 [INFO] About to spawn test binary
2024-06-26T11:52:01-04:00 [WARN] test binary failed to launch
2024-06-26T11:52:01-04:00 [WARN] test binary check failed
2024-06-26T11:52:01-04:00 [INFO] beginning to reinstall server
2024-06-26T11:52:01-04:00 [INFO] deleting server container
2024-06-26T11:52:01-04:00 [INFO] starting language server "package-version-server", path: "/Users/peter/source/robopaint", id: 3
2024-06-26T11:52:01-04:00 [INFO] starting language server. binary path: "/Users/peter/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.1/package-version-server", working directory: "/Users/peter/source/robopaint", args: []
2024-06-26T11:52:01-04:00 [ERROR] failed to start language server "package-version-server": failed to spawn command. path: "/Users/peter/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.1/package-version-server", working directory: "/Users/peter/source/robopaint", args: []
2024-06-26T11:52:01-04:00 [ERROR] server stderr: Some("")
2024-06-26T11:52:01-04:00 [INFO] retrying installation of language server "package-version-server" in 1s
2024-06-26T11:52:02-04:00 [ERROR] crates/languages/src/json.rs:393: No such file or directory (os error 2)
2024-06-26T11:52:02-04:00 [INFO] About to spawn test binary
2024-06-26T11:52:02-04:00 [WARN] test binary failed to launch
2024-06-26T11:52:02-04:00 [WARN] test binary check failed
2024-06-26T11:52:02-04:00 [INFO] beginning to reinstall server
2024-06-26T11:52:02-04:00 [INFO] deleting server container
2024-06-26T11:52:02-04:00 [ERROR] crates/language/src/language_registry.rs:880: server container removal
Caused by:
No such file or directory (os error 2)
2024-06-26T11:52:02-04:00 [INFO] starting language server "package-version-server", path: "/Users/peter/source/robopaint", id: 4
2024-06-26T11:52:02-04:00 [INFO] starting language server. binary path: "/Users/peter/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.1/package-version-server", working directory: "/Users/peter/source/robopaint", args: []
2024-06-26T11:52:02-04:00 [ERROR] failed to start language server "package-version-server": failed to spawn command. path: "/Users/peter/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.1/package-version-server", working directory: "/Users/peter/source/robopaint", args: []
2024-06-26T11:52:02-04:00 [ERROR] server stderr: Some("")
2024-06-26T11:52:02-04:00 [INFO] retrying installation of language server "package-version-server" in 1s
2024-06-26T11:52:04-04:00 [ERROR] crates/languages/src/json.rs:393: No such file or directory (os error 2)
2024-06-26T11:52:04-04:00 [INFO] About to spawn test binary
2024-06-26T11:52:04-04:00 [WARN] test binary failed to launch
2024-06-26T11:52:04-04:00 [WARN] test binary check failed
2024-06-26T11:52:04-04:00 [INFO] beginning to reinstall server
2024-06-26T11:52:04-04:00 [INFO] deleting server container
2024-06-26T11:52:04-04:00 [ERROR] crates/language/src/language_registry.rs:880: server container removal
</pre></details>

**Comment by ottob** [comment]
Thanks a lot for adding this, its super useful.

probably an edge case but we have entries like:
"native": "workspace:*",
in this case it tries to look for native package in npm, which is not correct.

**Comment by ocarreterom** [comment]
Hi @osiewicz 
The tooltip hides when I try to hover over it or move the cursor over the same package.

https://github.com/user-attachments/assets/a47ad130-214b-495f-aa0a-a2e233be29e5

**Comment by osiewicz** [comment]
Hey @ocarreterom, I believe this is an issue with our tooltips (e.g. https://github.com/zed-industries/zed/issues/13309) and not with the package.json dependency showing per se.

**Comment by osiewicz** [comment]
Actually, I take that back; the package language server does not fill out `range` property of the hover response, which may lead to the tooltip disappearing when the same package is being re-hovered.

**Comment by osiewicz** [comment]
@ocarreterom I've just released a new version of the package version server which sets the range in hover response; it feels way better to me. Give it a go and let me know how it goes. :) You don't need to update zed, but you may need to restart it for it to pick up updates to a language server.

**Comment by ocarreterom** [comment]
@osiewicz looks great! Thanks!

https://github.com/user-attachments/assets/2a0dd378-654f-4a3f-9161-513603448879

**Pull request merged**